### PR TITLE
Discord Bot API

### LINKED ISSUE #912: Discord Bot API
## Issue
We need an API to store the data of discord users who runs `/verify` command to verify themselves so that when verification process redirects to any Real Dev Squad website, we need to get the Discord ID of the user to link it successfully through the given link.

## What to do?
- [ ] Create a POST endpoint which can verify the connection using JWT between Cloudflare Worker and Backend and can also store the data in the backend in the below format
```
token: String,
discordId: String,
generationTime: Number,
expiry: Number,
linkStatus: Boolean,
```
**PR that creates a POST endpoint :** #911 

- [ ] Create a GET endpoint to get the Discord ID from generated token.

**PR that creates a GET endpoint :** TO BE CREATED

### PR DESCRIPTION
**Pull request by rohan09-raj**
## Issue
Resolves #912 

## Additions/Changes
- Made an API named as `/discord` which performs the action of adding the data coming from the `/verify` command 
- The server to server connection is secured by a JWT verification  from the headers

## API Contract
https://github.com/Real-Dev-Squad/website-api-contracts/pull/105/

## Data Model
https://github.com/Real-Dev-Squad/website-data-models/pull/19

### CODE REVIEW

File: config/development.js [hunk 1]
```diff
@@ -50,6 +50,47 @@ module.exports = {
...
+      "-----BEGIN PUBLIC KEY-----\n" +
+      "MIIBITANBgkqhkiG9w0BAQEFAAOCAQ4AMIIBCQKCAQBK3CkprcpAYxme7vtdjpWO\n" +
+      "gFFjoYsqU3OmhMEty/s1gnW5tgbK4ief4xk+cU+mu3YvjzWudT/SV17tAWxL4Y+G\n" +
+      "incJwL5gpQwlnw9qOAdRGkpBriQLec7kNVIydZXbUitziy+iSimxNzdDmjvlK9ZG\n" +
+      "miVLZm+MePbUtgaIpfgd+4bRWzudlITiNmWY7HppLzyBw+037iEICM4kwPPFI+SO\n" +
+      "GJhpAAmD6vk0MeZk1NeQmyQp/uOPpWmVRzgyK+XVc6AwZHV+/n6xAIT91/DjJlD1\n" +
+      "N+nS7Sqo3RJ04+KlNRUclzINOC7JBYkKtG7YQ0U9nNLkRrRlON+O6tY4OT86T1O1\n" +
+      "AgMBAAE=\n" +
+      "-----END PUBLIC KEY-----",
+    privateKey:
```
Comment (RitikJaiswal75) [inline config/development.js:64]: We should not give a private key over here as our backend does not require it.

File: config/test.js [hunk 2]
```diff
@@ -61,4 +61,45 @@ module.exports = {
       "Zzn584h/xbSIcP9/4SXmEK3wtoueUtMkvS3yKTvk+w==\n" +
       "-----END RSA PRIVATE KEY-----",
   },
+
+  botToken: {
+    publicKey:
```
Comment (RitikJaiswal75) [inline config/test.js:66]: are these keys different or the same as the above, if they are the same should we keep them here?

File: controllers/discord.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+const discordQuery = require("../models/discord");
```
Comment (RitikJaiswal75) [inline controllers/discord.js:1]: can we rename it to discordModel

File: services/botVerificationService.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+};
+
+/**
+ * Verifies if the JWT is valid. Throws error in case of signature error or expiry
+ *
+ * @param token {String} - JWT to be verified
+ * @return {Object} - Decode value of JWT
+ */
+const verifyToken = (token) => {
+  return jwt.verify(token, config.get("botToken.publicKey"), { algorithms: ["RS256"] });
```
Comment (RitikJaiswal75) [inline services/botVerificationService.js:23]: why aren't we checking the expiry here?

File: services/botVerificationService.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+const jwt = require("jsonwebtoken");
+const { CLOUDFLARE_WORKER } = require("../constants/bot");
+
+/**
+ * Generates the JWT
+ *
+ * @param payload {Object} - Payload to be added in the JWT
+ * @return {String} - Generated JWT
+ */
+const generateToken = () => {
```
Comment (RitikJaiswal75) [inline services/botVerificationService.js:10]: why are we generating a token here, token will be generated in Cloudflare worker right?

File: middlewares/authorizeBot.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const botVerifcation = require("../services/botVerificationService");
+const { CLOUDFLARE_WORKER } = require("../constants/bot");
+
+module.exports = async (req, res, next) => {
+  try {
+    const token = req.headers.authorization.split(" ")[1];
+    const data = botVerifcation.verifyToken(token);
```
Comment (RitikJaiswal75) [inline middlewares/authorizeBot.js:7]: why arent we checking the expiry of JWT here?

File: test/unit/models/discord.test.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  afterEach(async function () {
+    await cleanDb();
+  });
+
+  describe("addDiscordData", function () {
+    it("should add discord data to discord firestore collection", async function () {
+      const response = await discordQuery.addDiscordData(discordData[0]);
+
+      expect(response).to.be.an("object");
+      expect(response.message).to.equal("Added data successfully");
```
Comment (RitikJaiswal75) [inline test/unit/models/discord.test.js:18]: let's check for the fields as well in the test.

File: test/integration/discord.test.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
+const chai = require("chai");
+const { expect } = chai;
```
Comment (RitikJaiswal75) [inline test/integration/discord.test.js:2]: did we have sinon preinstalled or you installed it?

File: test/unit/middlewares/authorizeBot.test.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
+const authorizeBot = require("../../../middlewares/authorizeBot");
+const sinon = require("sinon");
```
Comment (RitikJaiswal75) [inline test/unit/middlewares/authorizeBot.test.js:2]: Why are we using sinon?

File: config/development.js (hunk 1 above)
Comment (rohan09-raj) [reply config/development.js:64]: We need it for testing purposes. Though I have removed the private key from here as it's not needed here but in `test.js` config

File: config/test.js (hunk 2 above)
Comment (rohan09-raj) [reply config/test.js:66]: They are same and we don't need distinct ones for development and testing purposes. We can keep them here as our production keys will be different from these, also we have done the same for `userToken` as well

File: controllers/discord.js (hunk 3 above)
Comment (rohan09-raj) [reply controllers/discord.js:1]: Done with the change

File: services/botVerificationService.js (hunk 5 above)
Comment (rohan09-raj) [reply services/botVerificationService.js:10]: Yes but for testing purposes we require a util which can be used anywhere in the testing part. Apart from that it's not used anywhere in the actual code

File: test/unit/models/discord.test.js (hunk 7 above)
Comment (rohan09-raj) [reply test/unit/models/discord.test.js:18]: It doesn't contain anything else other than `message`
image

File: test/unit/middlewares/authorizeBot.test.js (hunk 9 above)
Comment (rohan09-raj) [reply test/unit/middlewares/authorizeBot.test.js:2]: Sinon helps us for spying, mocking and stubbing. We are using it here for spying purposes to check the middleware output based on the inputs.

We can use a custom function as well, but that generally doesn't handle the cases and it is preinstalled so thought of using it

File: test/integration/discord.test.js (hunk 8 above)
Comment (rohan09-raj) [reply test/integration/discord.test.js:2]: It was preinstalled

File: services/botVerificationService.js (hunk 4 above)
Comment (rohan09-raj) [reply services/botVerificationService.js:23]: It is handled by the try catch where it is used

File: middlewares/authorizeBot.js (hunk 6 above)
Comment (rohan09-raj) [reply middlewares/authorizeBot.js:7]: So `jwt.verify` doesn't verify the expired token and ultimately throws an error. Therefore, the control goes to `catch` block and there we send the response of `Invalid Request`.
Do we want the user of this API to know that the token is expired?
My POV is we can do it for the developers to get better errors and debug the code faster if incase any bug arises.

File: test/unit/middlewares/authorizeBot.test.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    authorizeBot(request, response, nextSpy);
+    expect(nextSpy.calledOnce).to.be.equal(false);
+  });
+
+  it("return true when token is valid", function () {
+    const jwtToken = botVerifcation.generateToken();
+
+    const request = {
+      headers: {
+        authorization: `Bot ${jwtToken}`,
```
Comment (rohan09-raj) [inline test/unit/middlewares/authorizeBot.test.js:42]: I think we should be using the known authentication schemes here @RitikJaiswal75 
Check here : https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#authentication_schemes

Comment (rohan09-raj) [reply test/unit/middlewares/authorizeBot.test.js:42]: We can use `Bearer`

File: services/botVerificationService.js (hunk 4 above)
Comment (RitikJaiswal75) [reply services/botVerificationService.js:23]: JWT verify method throws an exception when the token is expired so try and catch can handle it.

File: services/botVerificationService.js (hunk 5 above)
Comment (RitikJaiswal75) [reply services/botVerificationService.js:10]: cant we directly write it as a function in test fixtures?

Not sure if it is a good approach please ask others about it as well

File: middlewares/authorizeBot.js (hunk 6 above)
Comment (RitikJaiswal75) [reply middlewares/authorizeBot.js:7]: ok
Would be easy if you could show a demo video.

File: test/unit/models/discord.test.js (hunk 7 above)
Comment (RitikJaiswal75) [reply test/unit/models/discord.test.js:18]: ok

File: test/unit/middlewares/authorizeBot.test.js (hunk 10 above)
Comment (RitikJaiswal75) [reply test/unit/middlewares/authorizeBot.test.js]: I am ok with anything here, I told bot as discord API's use `bot` 
Can you ask for others opinions as well?

File: test/unit/middlewares/authorizeBot.test.js (hunk 9 above)
Comment (RitikJaiswal75) [reply test/unit/middlewares/authorizeBot.test.js:2]: ok

File: config/development.js [hunk 11]
```diff
@@ -50,6 +50,19 @@ module.exports = {
       "-----END RSA PRIVATE KEY-----",
   },
 
+  botToken: {
+    publicKey:
```
Comment (RitikJaiswal75) [inline config/development.js:54]: won't we need this in prod?

Comment (prakashchoudhary07) [reply config/development.js:54]: We will be setting the keys from the hosting provider dashboard

Comment (rohan09-raj) [reply config/development.js:54]: Yeah right

File: config/default.js [hunk 12]
```diff
@@ -60,6 +60,11 @@ module.exports = {
     privateKey: "<privateKey>",
   },
 
+  botToken: {
+    publicKey: "<publicKey>",
+    privateKey: "<privateKey>",
+  },
+
```
Comment (prakashchoudhary07) [inline config/default.js:67]: Are these new keys that are being added from env?
Then we will have to add it in custom-environment-variables.js file

File: test/unit/middlewares/authorizeBot.test.js (hunk 10 above)
Comment (rohan09-raj) [reply test/unit/middlewares/authorizeBot.test.js]: Yeah sure

File: services/botVerificationService.js (hunk 5 above)
Comment (rohan09-raj) [reply services/botVerificationService.js:10]: We can create a util in testing helpers

File: config/default.js (hunk 12 above)
Comment (rohan09-raj) [reply config/default.js:67]: Added

File: services/botVerificationService.js (hunk 5 above)
Comment (rohan09-raj) [reply services/botVerificationService.js]: It's done

File: middlewares/authorizeBot.js (hunk 6 above)
Comment (rohan09-raj) [reply middlewares/authorizeBot.js:7]: [Screencast from 2023-01-31 19-55-13.webm](https://user-images.githubusercontent.com/78433013/215787664-d23ab69e-9fa8-423b-80c9-5de3e767a68a.webm)

Comment (RitikJaiswal75) [reply middlewares/authorizeBot.js:7]: but in our cfw we are setting expiry in jwt header
what about that scenario?

File: config/custom-environment-variables.js [hunk 13]
```diff
@@ -58,6 +58,10 @@ module.exports = {
     privateKey: "PRIVATE_KEY",
   },
 
+  botToken: {
+    publicKey: "PUBLIC_KEY",
```
Comment (prakashchoudhary07) [inline config/custom-environment-variables.js:62]: Can we change the name here, can we be more specific
And there already a key present with the same name

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: config/custom-environment-variables.js (hunk 13 above)
Comment (rohan09-raj) [reply config/custom-environment-variables.js]: Made the change

File: routes/discord.js [hunk 14]
```diff
@@ -0,0 +1,9 @@
+const express = require("express");
+const router = express.Router();
+const authorizeBot = require("../middlewares/authorizeBot");
+const validator = require("../middlewares/validators/discord");
+const discord = require("../controllers/discord");
+
+router.put("/", validator.discordData, authorizeBot, discord.addDiscordData);
```
[the remark this answers is not in the document]
Comment (rohan09-raj) [reply routes/discord.js:7]: There isn't any major reason for using `PUT` here but I have used it here as it is idempotent and duplicate resources won't be allowed and if we encounter any same resource it will update that the existing one.
So, we require this here as when the user initiates `/link` command it may happen that the bot server doesn't respond and user retries with same resource so in that case it will be helpful.

But `POST` will also work fine here as we will be handling the `Not responding` in Bot where it is meant to be used.

[the remark this answers is not in the document]
Comment (rohan09-raj) [reply routes/discord.js]: I have changed it to `POST`

File: config/development.js [hunk 15]
```diff
@@ -50,6 +50,19 @@ module.exports = {
       "-----END RSA PRIVATE KEY-----",
   },
 
+  botToken: {
+    botPublicKey:
+      "-----BEGIN PUBLIC KEY-----\n" +
```
Comment (RitikJaiswal75) [inline config/development.js:55]: Do you know if we need this in development?

Because without a private key, this public key is useless, 

Our public key will be saved in the discord bot, in development, we can update the readme and ask the developers to get their own public and private key pair and test it.

File: middlewares/validators/user.js [hunk 16]
```diff
@@ -106,6 +107,9 @@ async function getUsers(req, res, next) {
           "string.empty": "page must contain a positive number or zero",
           "string.pattern.base": "page value either be a positive number or zero",
         }),
+      search: joi.string().optional().messages({
```
Comment (RitikJaiswal75) [inline middlewares/validators/user.js:110]: is it a part of this PR? why do we need this?

File: routes/discord.js [hunk 17]
```diff
@@ -0,0 +1,9 @@
+const express = require("express");
+const router = express.Router();
+const authorizeBot = require("../middlewares/authorizeBot");
+const validator = require("../middlewares/validators/discord");
+const discord = require("../controllers/discord");
+
+router.post("/", validator.discordData, authorizeBot, discord.addDiscordData);
+
```
Comment (RitikJaiswal75) [inline routes/discord.js:8]: add 1 more route to GET the id with the matching token

File: test/integration/discord.test.js [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+          expect(res.body).to.have.property("message");
+          expect(res.body).to.have.property("error");
+          expect(res.body.message).to.equal("Unauthorized Bot");
+          expect(res.body.error).to.equal("Unauthorized");
+
+          return done();
+        });
+    });
+  });
+});
```
Comment (RitikJaiswal75) [inline test/integration/discord.test.js:104]: add a test for condition if the data is already present.

File: test/integration/users.test.js [hunk 19]
```diff
@@ -469,7 +469,7 @@ describe("Users", function () {
           return done();
         });
     });
-    it("Should return all users for empty value of search param", function (done) {
+    it("Should return 400 for empty value of search param", function (done) {
```
Comment (RitikJaiswal75) [inline test/integration/users.test.js:472]: this looks like a hot fix and it is not a part of discord API PR.

Please improve the test and can it be raised as a separate pr, we can get that merged before our discord API pr

**Review by RitikJaiswal75 — CHANGES_REQUESTED** [review]

File: config/test.js [hunk 20]
```diff
@@ -61,4 +61,45 @@ module.exports = {
...
+      "-----BEGIN PUBLIC KEY-----\n" +
+      "MIIBITANBgkqhkiG9w0BAQEFAAOCAQ4AMIIBCQKCAQBK3CkprcpAYxme7vtdjpWO\n" +
+      "gFFjoYsqU3OmhMEty/s1gnW5tgbK4ief4xk+cU+mu3YvjzWudT/SV17tAWxL4Y+G\n" +
+      "incJwL5gpQwlnw9qOAdRGkpBriQLec7kNVIydZXbUitziy+iSimxNzdDmjvlK9ZG\n" +
+      "miVLZm+MePbUtgaIpfgd+4bRWzudlITiNmWY7HppLzyBw+037iEICM4kwPPFI+SO\n" +
+      "GJhpAAmD6vk0MeZk1NeQmyQp/uOPpWmVRzgyK+XVc6AwZHV+/n6xAIT91/DjJlD1\n" +
+      "N+nS7Sqo3RJ04+KlNRUclzINOC7JBYkKtG7YQ0U9nNLkRrRlON+O6tY4OT86T1O1\n" +
+      "AgMBAAE=\n" +
+      "-----END PUBLIC KEY-----",
+    botPrivateKey:
```
Comment (harshith-venkatesh) [inline config/test.js:76]: Why are we storing confidential information

**Review by heyrandhir — COMMENTED** [review]
https://github.com/Real-Dev-Squad/website-backend/pull/911/files#r1096513937 . Good to go once we address comments by @harshith-venkatesh

File: config/development.js (hunk 15 above)
Comment (rohan09-raj) [reply config/development.js:55]: Yeah make sense but it would be easier if we direct them to use the privateKey from tests to generate the jwt token and developers can easily pass that jwt token to headers to test the API.

Developers can make mistakes as adding the key requires formatting and this will make them reach out to people to fix the bug. If we provide them formatted keys, they only need copy paste it and make the things work.

File: middlewares/validators/user.js (hunk 16 above)
Comment (rohan09-raj) [reply middlewares/validators/user.js]: Resolved

File: test/integration/discord.test.js (hunk 18 above)
Comment (rohan09-raj) [reply test/integration/discord.test.js:104]: This can't be tested as the data will be unique everytime because of the timestamps.

File: test/integration/users.test.js (hunk 19 above)
Comment (rohan09-raj) [reply test/integration/users.test.js]: Resolved

File: config/test.js (hunk 20 above)
Comment (rohan09-raj) [reply config/test.js:76]: This is only for testing purposes and without the dummy keys, we won't be able to test the API. The same is being done for `userToken` also.

File: routes/discord.js (hunk 17 above)
Comment (rohan09-raj) [reply routes/discord.js:8]: Will raise a separate PR for that, this PR has already lot of changes.

**Review by heyrandhir — APPROVED** [review]
LGTM

File: routes/discord.js (hunk 17 above)
Comment (RitikJaiswal75) [reply routes/discord.js:8]: ok please create a ticket and link it here so that it is not forgotten

File: config/test.js (hunk 20 above)
Comment (harshith-venkatesh) [reply config/test.js:76]: But Isn't this some account data that is registered, by providing this public.... Does it lead to anyone accessing ?

Comment (rohan09-raj) [reply config/test.js:76]: No, it's not registered with any account and do not lead any data access. I have created it for testing purposes from openssl.

Comment (harshith-venkatesh) [reply config/test.js:76]: Got it, then great :D

Comment (RitikJaiswal75) [reply config/test.js:76]: @harshith-venkatesh the original application is using the keys from github secrets. So rest assured, we are using a fresh set of keys for prod and staging.

File: constants/bot.js [hunk 21]
```diff
@@ -0,0 +1,4 @@
+const CLOUDFLARE_WORKER = "Cloudflare Worker";
+const BAD_TOKEN = "BAD.JWT.TOKEN";
```
[the remark this answers is not in the document]
Comment (harshith-venkatesh) [reply constants/bot.js:2]: Shouldn't this be handled as well?
Or are we creating a new PR to address this?

File: routes/discord.js (hunk 17 above)
Comment (rohan09-raj) [reply routes/discord.js:8]: Created a task in the same issue and updated the description
https://github.com/Real-Dev-Squad/website-backend/issues/912#issue-1557035449

**Review by harshith-venkatesh — APPROVED** [review]

**Review by RitikJaiswal75 — APPROVED** [review]

**Review by yesyash — APPROVED** [review]

File: test/integration/discord.test.js [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+          expect(res).to.have.status(401);
+          expect(res.body).to.have.property("message");
+          expect(res.body).to.have.property("error");
+          expect(res.body.message).to.equal("Unauthorized Bot");
+          expect(res.body.error).to.equal("Unauthorized");
+
+          return done();
+        });
+    });
```
Comment (rohan09-raj) [inline test/integration/discord.test.js:102]: @harshith-venkatesh

File: constants/bot.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (rohan09-raj) [reply constants/bot.js:2]: Actually I needed a BAD JWT Token to test the API therefore the JWT format was required to make the test pass, I checked for using other dummy values but it doesn't work.

Do you have any idea how I can test this?
Code Reference where it is used.
https://github.com/Real-Dev-Squad/website-backend/pull/911/files#r1096936950
